drivers/flash/flash_cadence_qspi_nor.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <device.h>
+
+#define DT_DRV_COMPAT cdns_qspi_nor
+
+#include <string.h>
+#include <drivers/flash.h>
+#include <init.h>
+#include <zephyr/logging/log.h>
+
+#include "flash_cadence_qspi_nor_ll.h"
+LOG_MODULE_REGISTER(flash_cadence, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_cad_priv {
+	DEVICE_MMIO_NAMED_RAM(qspi_reg);
+	DEVICE_MMIO_NAMED_RAM(qspi_data);
+	struct cad_qspi_params params;
+};
+
+struct flash_cad_config {
+	DEVICE_MMIO_NAMED_ROM(qspi_reg);
+	DEVICE_MMIO_NAMED_ROM(qspi_data);
+};
+
+static const struct flash_parameters flash_cad_parameters = {
+	.write_block_size = QSPI_BYTES_PER_DEV,
+	.erase_value = 0xff,
+};
+
+#define DEV_DATA(dev)	((struct flash_cad_priv *)((dev)->data))
+#define DEV_CFG(dev)	((struct flash_cad_config *)((dev)->config))
+
+static int flash_cad_read(const struct device *dev, off_t offset,
+				void *data, size_t len)
+{
+	int rc;
+
+	if (dev == NULL) {
+		LOG_ERR("No device detected");
+		return -ENODEV;
+	}
```

Comment:
```suggestion
```

this and others: it is redundant, `dev->api` is de-referenced before reaching this call, so a NULL device will lead to a crash before.

Reply:
This check is needed per our software development life cycle process, we need to make sure the pointer passed into the function is always checked. Without the checking, I will not able to pass our internal review, and upsteamed.

Reply:
> This check is needed per our software development life cycle process, we need to make sure the pointer passed into the function is always checked. Without the checking, I will not able to pass our internal review, and upsteamed.

As said: you'll never reach this point with dev == NULL, it will fail before at the API layer. If your development life cycle process depends on arg validation, you should instead request to modify APIs to do that and provide a rationale for that. Anyway, I don't think Zephyr needs to add runtime overhead for programming errors. Maybe asserts that at least can be disabled.